[OpenCL] Fixes adam_test fail for SYCL

jwlawson opened the pull request:
Adds and registers missing ops needed to pass the //tensorflow/python:adam_test test case.

I'm not sure if the `UnsortedSegmentSumFunctor` in  `tensorflow/core/kernels/segment_reduction_ops.cc` should be a templated version of the CPU version, as the implementation is exactly the same.

jwlawson committed:
Adds SYCL version of UnsortedSegmentSum

Uses a copy of the CPU version to provide an initial SYCL implementation
of the kernel. Doesn't add the other Segment ops which are included in
this file.

jwlawson committed:
Adds and registers SYCL version of GatherOp

Needed to change build file to include a dependency for gather_functor
on full tensorflow framework, rather than framework-lite.

The GatherOp uses the GatherFunctor, which is similar to the CPU
version, only the fast memcpy has been removed, so that Eigen handles
everything.

jwlawson committed:
Registers int64 SYCL version of Reshape op

Needed for the AdamOptimizer.

jwlawson committed:
Puts UnsortedSegmentSum segment_ids in host memory

The way the current kernel works on SYCL devices requires the
segment_ids to be accessed by the host, and so need to be available in
the host memory.

lukeiwanski commented on tensorflow/core/kernels/gather_functor.cc:
```diff
@@ -43,12 +46,44 @@ TF_CALL_complex128(DECLARE_GPU_SPECS);
 
-#include "tensorflow/core/kernels/gather_functor.h"
+TF_CALL_GPU_NUMBER_TYPES_NO_HALF(DECLARE_GPU_SPECS);
+// TF_CALL_complex64(DECLARE_GPU_SPECS);
```
Remove commented out code

lukeiwanski commented on tensorflow/core/kernels/gather_functor.cc:
```diff
@@ -43,12 +46,44 @@ TF_CALL_complex128(DECLARE_GPU_SPECS);
-#include "tensorflow/core/kernels/gather_functor.h"
+TF_CALL_GPU_NUMBER_TYPES_NO_HALF(DECLARE_GPU_SPECS);
+// TF_CALL_complex64(DECLARE_GPU_SPECS);
+// TF_CALL_complex128(DECLARE_GPU_SPECS);
```
Remove commented out code

lukeiwanski commented on tensorflow/core/kernels/BUILD:
```diff
@@ -993,7 +993,8 @@ tf_kernel_library(
     visibility = [":friends"],
     deps = [
         ":bounds_check",
-        "//tensorflow/core:framework_lite",
```
we probably should not remove that.
Can we wrap our dependency in if_sycl() ? Like done here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/BUILD#L498 
Or is that needed on CPU side as well?

lukeiwanski reviewed (changes requested):
Overall great! Couple minor comments

jwlawson committed:
Removes unwanted commented out code

jwlawson committed:
Only require full framework for SYCL, not CPU

The register_types.h header (providing TF_CALL_...) pulls in some
protobuf headers, which are only provided by the full framework, not the
framework_lite. Add this dependency when building with SYCL support,
but not when building for CPU only.

lukeiwanski commented on tensorflow/core/kernels/BUILD:
```diff
@@ -995,7 +995,9 @@ tf_kernel_library(
         "//tensorflow/core:framework_lite",
         "//third_party/eigen3",
-    ],
+    ] + if_sycl([
```
Instead of doing that can we add 
`if_sycl(["//tensorflow/core:sycl_runtime"])` to framework_lite ? ( https://github.com/lukeiwanski/tensorflow/blob/master/tensorflow/core/BUILD#L471 )

jwlawson replied:
We cannot do this, as `:sycl_runtime` itself depends on the full `:framework`, so this would force `:framework_lite` to depend on `:framework`.

lukeiwanski replied:
ah! OK makes sense! Thanks.

keryell commented on tensorflow/core/kernels/gather_functor.cc:
```diff
@@ -43,12 +46,44 @@ TF_CALL_complex128(DECLARE_GPU_SPECS);
-}  // namespace functor
-}  // namespace tensorflow
+#ifdef TENSORFLOW_USE_SYCL
+// Stipped down version of CPU functor which uses Eigen loops to copy slices
```
Stripped

keryell commented on tensorflow/core/kernels/segment_reduction_ops.cc:
```diff
@@ -413,6 +442,24 @@ TF_CALL_complex128(REGISTER_GPU_UNSORTED_KERNELS_ALL);
+#if TENSORFLOW_USE_SYCL
+#define REGISTER_GPU_UNSORTED_KERNELS(type, index_type)                \
+  REGISTER_KERNEL_BUILDER(Name("UnsortedSegmentSum")                   \
+                              .Device(DEVICE_SYCL)                      \
```
Alignment of `\`

keryell reviewed (changes requested):
Thank you for the contribution.

Just a few typos on my side.

Be sure to update `sycl` to `SYCL` in the mErgE RequESt mESSage wHEn iT Is mERgED :-)

jwlawson committed:
Fixes spelling

lukeiwanski commented:
@keryell  are you happy with the changes?
